http-clients/apache5-client/src/main/java/software/amazon/awssdk/http/apache5/Apache5HttpClient.java

```diff
@@ -250,20 +250,28 @@ public void abort() {
     public void close() {
         HttpClientConnectionManager cm = httpClient.getHttpClientConnectionManager();
         IdleConnectionReaper.getInstance().deregisterConnectionManager(cm);
-        cm.shutdown();
+        // TODO : need to add test cases for this
+        cm.close(CloseMode.IMMEDIATE);
     }
 
     private HttpExecuteResponse execute(HttpUriRequestBase apacheRequest, MetricCollector metricCollector) throws IOException {
         HttpClientContext localRequestContext = Apache5Utils.newClientContext(requestConfig.proxyConfiguration());
         THREAD_LOCAL_REQUEST_METRIC_COLLECTOR.set(metricCollector);
         try {
-            HttpResponse httpResponse = httpClient.execute(apacheRequest, localRequestContext);
-            return createResponse(httpResponse, apacheRequest);
+            return httpClient.execute(apacheRequest, localRequestContext, response -> {
+
+                // TODO : This is required since Apache5 closes streams immediately, check memory impacts because of this.
+                if (response.getEntity() != null) {
+                    response.setEntity(new BufferedHttpEntity(response.getEntity()));
```

Comment:
Hmm, this seems odd. We should not buffer the entire stream.

Reply:
Yeap I agree , but as mentioned in 
```
The use of an HTTP response handler guarantees that the underlying HTTP connection will be released back to the connection manager automatically in all cases.
```

https://hc.apache.org/httpcomponents-client-5.4.x/examples.html

Thus we need to buffer the stream , we need to find a way around this or some better approach.